packages/completer/src/handler.ts

```diff
@@ -474,7 +536,82 @@ export namespace CompletionHandler {
      * it is acceptable for the other methods to be simple functions that return
      * rejected promises.
      */
-    connector: IDataConnector<IReply, void, IRequest>;
+    connector?: IDataConnector<IReply, void, IRequest>;
```

Comment:
@afshin and I had a good discussion about this in the meeting today. My takeaway was that we should allow you to specify either type of connector, and then after you call fetch, check the return type to see which type was provided, by looking at the keys provided in the object 🤷.

Maybe with the idea that we would deprecate the first option eventually? 

```suggestion
    connector: IDataConnector<IReply, void, IRequest> | IDataConnector<ICompletionItemsReply, void, IRequest>;;
```

Reply:
I like it! I'll go ahead and make this change.

Reply:
On this topic, I'm using type assertions to differentiate the different replies, but it seems kind of messy. Do you guys have any guidance on the best practice here? ([Original implementation](https://github.com/jupyterlab/jupyterlab/blob/380ad1a4ce8ba0f5a7b0490f3119f498cae776ae/packages/completer/src/handler.ts#L359) for reference)

```typescript
    const response = this._connector.fetch(request);
    return (response as Promise<
      CompletionHandler.ICompletionItemsReply | undefined
    >)
      .then(reply => {
      // ...
        if (!reply.items) {
          throw new Error(
            'Casting response as Promise<CompletionHandler.IReply'
          );
        }

        this._onFetchItemsReply(state, reply);
      })
      .catch(_ => {
        (response as Promise<CompletionHandler.IReply | undefined>)
          .then(reply => {
               //...
               this._onReply(state, reply);
          })
          .catch(_ => this._onFailure());
      });
```

Reply:
Yeah it's messy... But I think you are doing it right. If you used the tag union approach like you mentioned below it might be a bit cleaner?